pkg/cmd/httpd/server.go

```
@@ -0,0 +1,292 @@
+			return
+		}
+
+		cmd := exec.Command(cmd, append(args, req.Args...)...)
+		cmd.Env = os.Environ()
+		for k, v := range req.Env {
+			cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", k, v))
+		}
+		stdoutPipe, err := cmd.StdoutPipe()
+		if err != nil {
+			WriteHTTPError(w, r, err)
+			return
+		}
+		stderrPipe, err := cmd.StderrPipe()
+		if err != nil {
+			WriteHTTPError(w, r, err)
+			return
+		}
+
+		cmdExecutorsMutex.Lock()
+		executionID := uuid.NewString()
+		cmdExecutor := GetCmdExecutor()
+		cmdExecutors[executionID] = cmdExecutor
+		cmdExecutorsMutex.Unlock()
+		if err := cmdExecutor.Execute(cmd); err != nil {
+			WriteHTTPError(w, r, err)
+			return
+		}
+
+		flusher, ok := w.(http.Flusher)
+		if !ok {
+			WriteHTTPError(w, r, errors.New("expected http.ResponseWriter to be an http.Flusher"))
+			return
+		}
+
+		// After setting these headers, we can't call `WriteHTTPError` anymore to write an error.
+		// Instead, we should write the error as a StatusOutput message.
+		w.Header().Set("Connection", "Keep-Alive")
+		w.Header().Set("X-Content-Type-Options", "nosniff")
+		w.Header().Set("Transfer-Encoding", "chunked")
+
+		outputC := make(chan Output, 4096)
+		errC := make(chan error)
+		outputDoneC := make(chan interface{}, 2)
+
+		// Write an output so the client knows the executionID.
+		outputC <- Output{
+			Msg:         "",
+			Type:        OutputTypeStdout,
+			ExecutionID: executionID,
+		}
+		go WriteOutput(stdoutPipe, OutputTypeStdout, outputDoneC, outputC, errC, executionID)
+		go WriteOutput(stderrPipe, OutputTypeStderr, outputDoneC, outputC, errC, executionID)
+		go func() {
+			for {
+				select {
+				case <-r.Context().Done():
+					logger.Log("context finished")
+					if r.Context().Err() != nil {
+						errC <- r.Context().Err()
+					}
+					return
+				}
+			}
+		}()
+
+		if err := cmdExecutor.Run(outputC, outputDoneC, errC, &ChunkedEncoder{
+			Encoder: json.NewEncoder(w),
+			Flusher: flusher,
+		}); err != nil {
```

Let's log this error too

---

Actually, can we return these errors via http too? So that our runner can deal with them